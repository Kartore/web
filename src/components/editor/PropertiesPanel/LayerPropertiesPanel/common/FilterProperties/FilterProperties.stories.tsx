import { FilterProperties } from '.';
import { Meta, StoryObj } from '@storybook/react';

const meta = {
  component: FilterProperties,
} satisfies Meta<typeof FilterProperties>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {},
};
