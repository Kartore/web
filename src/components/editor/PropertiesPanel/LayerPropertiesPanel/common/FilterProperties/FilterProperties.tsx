import type { ComponentProps, FC } from 'react';

import type {
	BackgroundLayerSpecification,
	LayerSpecification,
} from '@maplibre/maplibre-gl-style-spec';

import { FilterInputField } from '~/components/common/FilterInputField';
import { MonacoEditor } from '~/components/common/MonacoEditor';
import { isExpressionFilter } from '~/components/editor/PropertiesPanel/LayerPropertiesPanel/common/FilterProperties/utils/isExpressionFilter.ts';
import { getStyleJSONSchemaDefinition } from '~/components/editor/PropertiesPanel/LayerPropertiesPanel/common/RawDataProperties/schema/StyleJSONSchemaBase.ts';
import type { onChangeType } from '~/components/editor/PropertiesPanel/LayerPropertiesPanel/utils/LayerUtil/LayerUtil.ts';
import { cn } from '~/utils/tailwindUtil.ts';

export type FilterPropertiesProps = Omit<ComponentProps<'div'>, 'onChange'> & {
	layer: Exclude<LayerSpecification, BackgroundLayerSpecification>;
	onChange?: onChangeType;
};

export const FilterProperties: FC<FilterPropertiesProps> = ({
	layer,
	onChange,
	children,
	className,
	...props
}) => {
	return (
		<div {...props} className={cn('flex flex-col gap-2 px-4', className)}>
			<h3 className={'font-montserrat font-semibold text-sm'}>Filter</h3>
			{isExpressionFilter(layer.filter) ? (
				<FilterInputField
					value={layer.filter}
					onChange={(value) => onChange?.(layer, undefined, 'filter', value)}
				/>
			) : (
				<summary>
					<details>
						<MonacoEditor
							className={cn('min-h-16', className)}
							value={JSON.stringify(
								{ filter: layer.filter ? layer.filter : [] },
								undefined,
								2,
							)}
							onChange={(value) => {
								if (onChange && value) {
									const filterValue = JSON.parse(value).filter;
									onChange(
										layer,
										undefined,
										'filter',
										filterValue.length === 0 ? undefined : filterValue,
									);
								}
							}}
							onMount={(_, monaco) => {
								monaco.languages.json.jsonDefaults.setDiagnosticsOptions({
									validate: true,
									schemas: [
										{
											fileMatch: ['*'],
											uri: 'kartore://FilterSpecification.json',
											schema: {
												type: 'object',
												properties: {
													filter: getStyleJSONSchemaDefinition(
														'FilterSpecification',
													),
												},
											},
										},
									],
								});
							}}
						/>
					</details>
				</summary>
			)}
			{children}
		</div>
	);
};

FilterProperties.displayName = 'FilterProperties';
