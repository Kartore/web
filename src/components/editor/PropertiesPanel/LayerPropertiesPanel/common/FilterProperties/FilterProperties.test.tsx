import { describe } from 'vitest';
import { render } from '@testing-library/react';
import { composeStories } from '@storybook/react';

import * as Stories from './FilterProperties.stories';

const { Default } = composeStories(Stories);

describe('Component: FilterProperties', () => {
  describe('Snapshot', () => {
    it('Default', () => {
      const { asFragment } = render(<Default />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
