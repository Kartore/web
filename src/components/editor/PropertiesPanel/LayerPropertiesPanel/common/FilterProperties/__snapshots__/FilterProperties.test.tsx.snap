// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: FilterProperties > Snapshot > Default 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2 px-4"
  >
    <h3
      class="font-montserrat text-sm font-semibold"
    >
      Filter
    </h3>
    <section
      style="display: flex; position: relative; text-align: initial; width: 100%; height: 100%;"
    >
      <div
        style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
      >
        Loading...
      </div>
      <div
        class="min-h-16"
        style="width: 100%; display: none;"
      />
    </section>
  </div>
</DocumentFragment>
`;
